internals/plan/plan.go

```diff
@@ -1085,19 +1228,88 @@ func (p *Plan) checkCycles() error {
 }
 
 func ParseLayer(order int, label string, data []byte) (*Layer, error) {
-	layer := Layer{
-		Services:   map[string]*Service{},
-		Checks:     map[string]*Check{},
-		LogTargets: map[string]*LogTarget{},
-	}
-	dec := yaml.NewDecoder(bytes.NewBuffer(data))
-	dec.KnownFields(true)
-	err := dec.Decode(&layer)
+	layer := &Layer{
+		Services:   make(map[string]*Service),
+		Checks:     make(map[string]*Check),
+		LogTargets: make(map[string]*LogTarget),
+		Sections:   make(map[string]Section),
+	}
+
+	// The following manual approach is required because:
+	//
+	// 1. Extended sections are YAML inlined, and also do not have a
+	// concrete type at this level, we cannot simply unmarshal the layer
+	// in one step.
+	//
+	// 2. We honor KnownFields = true behaviour for non extended schema
+	// sections, and at the top field level, which includes Section field
+	// names.
+	builtins := map[string]interface{}{
+		"summary":     &layer.Summary,
+		"description": &layer.Description,
+		"services":    &layer.Services,
+		"checks":      &layer.Checks,
+		"log-targets": &layer.LogTargets,
+	}
+	// Make sure builtins contains the exact same fields as expected
+	// in the Layer type.
+	if !mapMatchKeys(builtins, builtinSections) {
```

Comment:
Is this something we even need to do at runtime, given it's a coding bug if we get it wrong? Or could we just have a test for it?

Reply:
I've removed the runtime check. Also, since it does not cover all the hard-coded field cases, and settled with an updated test which fails on changes to the exported YAML field in both Layer and Plan types.